mongodb-plugin/src/main/java/io/cdap/plugin/batch/source/MongoDBBatchSource.java

```diff
@@ -0,0 +1,176 @@
+                                       config.splitterClass);
+      Class<? extends MongoSplitter> klass = getClass().getClassLoader().loadClass(
+        className).asSubclass(MongoSplitter.class);
+      MongoConfigUtil.setSplitterClass(conf, klass);
+    }
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+    lineageRecorder.createExternalDataset(config.getSchema());
+    List<Schema.Field> fields = config.getSchema().getFields();
+    if (fields != null && !fields.isEmpty()) {
+      lineageRecorder.recordRead("Read", "Read from MongoDB collection.",
+                                 fields.stream().map(Schema.Field::getName).collect(Collectors.toList()));
+    }
+    context.setInput(Input.of(config.referenceName,
+                              new SourceInputFormatProvider(MongoConfigUtil.getInputFormat(conf), conf)));
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    bsonConverter = new BSONConverter(config.getSchema());
+  }
+
+  @Override
+  public void transform(KeyValue<Object, BSONObject> input, Emitter<StructuredRecord> emitter) throws Exception {
+    BSONObject bsonObject = input.getValue();
+    emitter.emit(bsonConverter.transform(bsonObject));
+  }
+
+  /**
+   * Config class for {@link MongoDBBatchSource}.
+   */
+  public static class MongoDBSourceConfig extends MongoDBConfig {
+
+    @Name(MongoDBConstants.SCHEMA)
+    @Description("Schema of records output by the source.")
+    private String schema;
+
+    @Name(MongoDBConstants.INPUT_QUERY)
+    @Description("Optionally filter the input collection with a query. This query must be represented in JSON " +
+      "format, and use the MongoDB extended JSON format to represent non-native JSON data types.")
+    @Nullable
+    @Macro
+    private String inputQuery;
+
+    @Name(MongoDBConstants.INPUT_FIELDS)
+    @Nullable
+    @Description("A projection document limiting the fields that appear in each document. " +
+      "If no projection document is provided, all fields will be read.")
+    @Macro
+    private String inputFields;
+
+    @Name(MongoDBConstants.SPLITTER_CLASS)
+    @Nullable
+    @Description("The name of the Splitter class to use. If left empty, the MongoDB Hadoop Connector will attempt " +
+      "to make a best guess as to what Splitter to use.")
+    @Macro
+    private String splitterClass;
+
+    @Name(MongoDBConstants.AUTH_CONNECTION_STRING)
+    @Nullable
+    @Description("Auxiliary MongoDB connection string to authenticate against when constructing splits.")
+    @Macro
+    private String authConnectionString;
+
+    /**
+     * @return {@link Schema} of the dataset if one was given else null
+     * @throws IllegalArgumentException if the schema is null or not a valid JSON
+     */
+    public Schema getSchema() {
+      if (schema == null) {
+        throw new IllegalArgumentException("Schema cannot be null.");
```

Comment:
this error type should not be thrown during validation phase. Should be InvalidConfigPropertyException. Also IAE usually thrown on illegal method param, in cases like this it looks more like IllegalStateException

Reply:
[Changed](https://github.com/vladhlinskiy/database-plugins/pull/14/files#diff-56e5c8c248fc520b3c862df33668d2feR178) to:
```
    /**
     * @return the schema of the dataset
     */
    @Nullable
    public Schema getSchema() {
      try {
        return null == schema ? null : Schema.parseJson(schema);
      } catch (IOException e) {
        throw new InvalidConfigPropertyException("Invalid schema", e, MongoDBConstants.SCHEMA);
      }
    }
```